daemon/cmd/settings.go

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	k8sConsts "github.com/cilium/cilium/pkg/k8s/constants"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/option/resolver"
+)
+
+// settingsDelta tracks deltas between agent settings and cilium-config configmap.
+type settingsDelta struct {
+	// deltas tracks deltas between ConfigMap and agent settings.
+	deltas []string
+	// md5 tracks configmap version
+	md5 string
+}
+
+func (s settingsDelta) hasDeltas() bool {
+	return len(s.deltas) != 0
+}
+
+func (s settingsDelta) deltaCount() int {
+	return len(s.deltas)
+}
+
+// newSettingsDelta computes deltas between daemon settings and cilium config map.
+// If per node config is available, overlay the overrides prior to computing the deltas.
+func newSettingsDelta(ctx context.Context, d *Daemon) (settingsDelta, error) {
+	var (
+		sd   settingsDelta
+		node = os.Getenv(k8sConsts.EnvNodeNameSpec)
+	)
+
+	desiredCfg, _, err := resolver.ReadConfigSource(ctx, d.clientset, node, resolver.ConfigSource{
+		Kind:      resolver.KindConfigMap,
+		Namespace: option.Config.K8sNamespace,
+		Name:      "cilium-config",
+	})
+	if err != nil {
+		return sd, err
+	}
+	noCfg, _, err := resolver.ReadConfigSource(ctx, d.clientset, node, resolver.ConfigSource{
+		Kind:      resolver.KindNodeConfig,
+		Namespace: option.Config.K8sNamespace,
+	})
+	if err != nil {
+		log.Debugf("cilium node configuration load failed: %s", err)
+	}
+	for k, v := range noCfg {
+		if _, ok := desiredCfg[k]; ok {
+			desiredCfg[k] = v
+		}
+	}
+	sd.deltas = computeSettingsDeltas(desiredCfg, d.settings)
```

Comment:
Both viper and the cell infrastructure do some pretty heavy preprocessing of the configmap values [e.g. here](https://github.com/cilium/cilium/blob/519d1c83ae676f82bac1b0e0e3df28bfedf31ad2/pkg/hive/cell/config.go#L107-L120). It seems that `d.settings` contains values before the hive preprocessing, but still went through the Viper [parser](https://github.com/cilium/cilium/blob/cb0ed6d5de219ef25ed4ac4d5a7d40fea560ab7c/pkg/hive/hive.go#L252) here. Given that our `desiredCfg` did no go through Viper, where as `d.settings` was parsed by Viper and then [converted back to a string](https://github.com/cilium/cilium/blob/61477fce8e713835f999409cd1ab292732729720/daemon/cmd/debuginfo.go#L72), are we confident that the string values are always equivalent?

Unfortunately I'm no longer familiar with how cilium-agent parses the flags in detail, but I think it is fair to say that it is pretty complex and subject to change. Maybe we should just fetch a raw snapshot of the ConfigMap mount (which has node overrides applied) before it goes through viper and hive and compare that snapshot to `desiredCfg`, rather than comparing to `d.settings`? The ConfigMap mount is read here: https://github.com/cilium/cilium/blob/6fbadc04b0dc2dfe93259a0388dfac0d530cf11b/pkg/option/config.go#L4173

Reply:
I'll change to read from the config dir directly as it seems a lesser evil?